src/lib/statsig.ts

```diff
@@ -1,20 +1,69 @@
-// import { StatsigClient } from '@statsig/js-client';
+import { StatSigFlags, StatsigConfigType } from '@/types/statsig';
+import { StatsigClient } from '@statsig/js-client';
+import { merge } from 'lodash';
 
-// let statsigClient: StatsigClient;
+import { log } from './telemetry';
 
-// export const initStatsig = async () => {
-//   if (statsigClient) return statsigClient;
-//   statsigClient = new StatsigClient(
-//     import.meta.env.VITE_STATSIG_CLIENT_KEY,
-//     {},
-//     {
-//       disableLogging: import.meta.env.VITE_DISABLE_STATSIG,
-//       disableStorage: import.meta.env.VITE_DISABLE_STATSIG,
-//     }
-//   );
-//   await statsigClient.initializeAsync();
-//   return statsigClient;
-// };
+let statsigClient: StatsigClient;
+let initPromise: Promise<StatsigClient> | null = null;
 
-// TODO: figure out how to init statsig async so we get configs on the first page load without breaking app state
-// https://linear.app/dydx/project/feature-experimentation-6853beb333d7/overview
+/**
+ * This method creates a promise and assigns it to the variable initPromise.
+ * If an initPromise has already been generated, it returns the existing promise.
+ * This causes all calls to this function to always return the statsigClient only after
+ * it has retrieved all values from statsig
+ *
+ * @returns initPromise: a promise that returns the statsig client only after it is initialized
+ */
+export const initStatsig = async () => {
+  if (initPromise) {
+    return initPromise;
+  }
+  if (statsigClient) return statsigClient;
+  initPromise = (async () => {
+    statsigClient = new StatsigClient(
+      import.meta.env.VITE_STATSIG_CLIENT_KEY ?? '',
+      // TODO: create a top level settings.ts file to coerce boolean env vars to actual boolean
+      import.meta.env.VITE_TEST_USER_ID === 'true' ? { userID: 'test-id' } : {},
+      {
+        disableLogging: import.meta.env.VITE_DISABLE_STATSIG,
+        disableStorage: import.meta.env.VITE_DISABLE_STATSIG,
+      }
+    );
+    await statsigClient.initializeAsync();
+    return statsigClient;
+  })();
+  return initPromise;
+};
+
+/**
+ *
+ * @param client statsig client to check
+ * @param gateId
+ * @returns
+ */
+const checkGateTyped = (client: StatsigClient, gateId: StatSigFlags) => {
+  return client.checkGate(gateId);
+};
+
+/**
+ *
+ * This is used only in useInitializePage to block abacus start on the fully loaded async statsig config.
+ * This prevents new users from automatically received a 'false' for all feature gates because they
+ * don't have a statsig value cached.
+ *
+ * Try catches in case the statsig init or gatechecks fail.
+ */
+export const statsigConfigPromise = async (): Promise<StatsigConfigType> => {
+  try {
+    const client = await initStatsig();
+    const [firstGateResults, ...otherGateResults] = Object.values(StatSigFlags).map((gateId) => ({
```

Comment:
@tyleroooo invert messes up the typing so it's hard to be certain we're using the correct gateId's so i instead used a map + merge. LMK if this works for you :)

Reply:
what about object.assign({}, allObjects)

Reply:
or merge ({}, allObjects)